src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/SqlImportOperation.cs

```diff
@@ -0,0 +1,388 @@
+            using (SqlBulkCopy bulkCopy = new SqlBulkCopy(sqlConnectionWrapper.SqlConnection, SqlBulkCopyOptions.CheckConstraints | SqlBulkCopyOptions.UseInternalTransaction | SqlBulkCopyOptions.KeepNulls, null))
+            {
+                bulkCopy.DestinationTableName = dataTable.TableName;
+                bulkCopy.BatchSize = dataTable.Rows.Count;
+
+                try
+                {
+                    await _sqlServerTransientFaultRetryPolicyFactory.Create().ExecuteAsync(
+                        async () =>
+                        {
+                            bulkCopy.BulkCopyTimeout = _importTaskConfiguration.SqlBulkOperationTimeoutInSec;
+                            await bulkCopy.WriteToServerAsync(dataTable.CreateDataReader());
+                        });
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogInformation(ex, "Failed to bulk copy data.");
+
+                    throw;
+                }
+            }
+        }
+
+        public async Task<IEnumerable<SqlBulkCopyDataWrapper>> BulkMergeResourceAsync(IEnumerable<SqlBulkCopyDataWrapper> resources, CancellationToken cancellationToken)
+        {
+            List<long> importedSurrogatedId = new List<long>();
+
+            // Make sure there's no dup in this batch
+            resources = resources.GroupBy(r => (r.ResourceTypeId, r.Resource.ResourceId)).Select(r => r.First());
+            IEnumerable<BulkImportResourceTypeV1Row> inputResources = resources.Select(r => r.BulkImportResource);
+
+            using (SqlConnectionWrapper sqlConnectionWrapper = await _sqlConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken, true))
+            using (SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand())
+            {
+                try
+                {
+                    VLatest.BulkMergeResource.PopulateCommand(sqlCommandWrapper, inputResources);
+                    sqlCommandWrapper.CommandTimeout = _importTaskConfiguration.SqlBulkOperationTimeoutInSec;
+
+                    var sqlDataReader = await sqlCommandWrapper.ExecuteReaderAsync(cancellationToken);
+
+                    while (await sqlDataReader.ReadAsync(cancellationToken))
+                    {
+                        long surrogatedId = sqlDataReader.GetInt64(0);
+                        importedSurrogatedId.Add(surrogatedId);
+                    }
+
+                    return resources.Where(r => importedSurrogatedId.Contains(r.ResourceSurrogateId));
+                }
+                catch (SqlException sqlEx)
+                {
+                    _logger.LogError(sqlEx, "Failed to merge resources. " + sqlEx.Message);
+
+                    throw;
+                }
+            }
+        }
+
+        public async Task CleanBatchResourceAsync(string resourceType, long beginSequenceId, long endSequenceId, CancellationToken cancellationToken)
```

Comment:
I've done some successful imports, but also some problems if the task is killed before the import is complete. For example:

```
fail: Microsoft.Health.Fhir.SqlServer.Features.Storage.SqlServerFhirModel[0]
      The SqlServerFhirModel instance has not run the initialization required for the current schema version
info: Microsoft.Health.Fhir.SqlServer.Features.Operations.Import.SqlResourceBulkImporter[0]
      Failed to clean batch resource.
      Microsoft.Health.Abstractions.Exceptions.ServiceUnavailableException: The operation could not be completed, because the service was unable to accept new requests. It is safe to retry the operation. If the issue persists, please contact support.
         at Microsoft.Health.Fhir.SqlServer.Features.Storage.SqlServerFhirModel.ThrowIfNotInitialized() in C:\src\fhir-server\src\Microsoft.Health.Fhir.SqlServer\Features\Storage\SqlServerFhirModel.cs:line 462
         at Microsoft.Health.Fhir.SqlServer.Features.Storage.SqlServerFhirModel.GetResourceTypeId(String resourceTypeName) in C:\src\fhir-server\src\Microsoft.Health.Fhir.SqlServer\Features\Storage\SqlServerFhirModel.cs:line 98
         at Microsoft.Health.Fhir.SqlServer.Features.Storage.SqlImportOperation.CleanBatchResourceAsync(String resourceType, Int64 beginSequenceId, Int64 endSequenceId, CancellationToken cancellationToken) in C:\src\fhir-server\src\Microsoft.Health.Fhir.SqlServer
\Features\Operations\Import\SqlImportOperation.cs:line 186
         at Microsoft.Health.Fhir.SqlServer.Features.Operations.Import.SqlResourceBulkImporter.CleanResourceAsync(ImportProcessingTaskInputData inputData, ImportProcessingProgress progress, CancellationToken cancellationToken) in C:\src\fhir-server\src\Microsoft.
Health.Fhir.SqlServer\Features\Operations\Import\SqlResourceBulkImporter.cs:line 155.
```

Reply:
Thanks for catching this, update with fix.